### PR TITLE
QUICKSTEP-127 Data provider thread 

### PR DESCRIPTION
**Pull request by hbdeshmukh**
- Thread used for receiving block requests from clients and sending the
contents of a block back to the client.

### CODE REVIEW

**Comment by hbdeshmukh** [comment]
@zuyu Question: How do you run tests for distributed quickstep on Travis?

File: storage/DataProviderThread.hpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+class QueryProcessor;
+class StorageManager;
+
+/** \addtogroup Storage
+ *  @{
+ */
+
+/**
+ * @brief A thread that provides access to query results (e.g. storage blocks),
+ *        to Quickstep clients.
```
Comment (zuyu) [inline storage/DataProviderThread.hpp]: Could you please provide more info regarding this class, including which class will own this class, and why it needs write permissions to both `CatalogDatabase` and `QueryProcessor`? Thanks!

**Comment by zuyu** [comment]
We do not test the distributed version in CI due to previous time-out issue while installing `grpc` by source code.

File: storage/DataProviderThread.hpp (hunk 1 above)
Comment (hbdeshmukh) [reply storage/DataProviderThread.hpp]: The DataProviderThread class will be part of the Quickstep server process. The following image describes the role of this thread. 
image

Comment (hbdeshmukh) [reply storage/DataProviderThread.hpp]: This thread will be responsible to drop temporary tables created by queries. To perform this drop operation, it needs access to both CatalogDatabase and QueryProcessor. Details in the code [here](https://github.com/hbdeshmukh/incubator-quickstep/blob/6967ca57c448a17b63e176bf8bc1576066fe893a/storage/DataProviderThread.cpp#L79)

**Comment by hbdeshmukh** [comment]
Right now, the new code is included only in the distributed version. It need not be the case, as the only dependency for Data Provider Thread is for GRPC and Networked TMB.

**Comment by hbdeshmukh** [comment]
Hi @zuyu I have made the changes. The code in this PR will be invoked only with a compile time flag ENABLE_NETWORK_CLI. Earlier it was dependent on ENABLE_DISTRIBUTED, which was not correct. I am working on interfacing this feature with our frontend, another PR will be coming soon which will have an end-to-end working of this feature.

File: storage/CMakeLists.txt [hunk 2]
```diff
@@ -120,6 +120,11 @@ configure_file (
 QS_PROTOBUF_GENERATE_CPP(storage_AggregationOperationState_proto_srcs
                          storage_AggregationOperationState_proto_hdrs
                          AggregationOperationState.proto)
+if (ENABLE_NETWORK_CLI)
+QS_PROTOBUF_GENERATE_CPP(storage_BlockWire_proto_srcs
```
Comment (zuyu) [inline storage/CMakeLists.txt]: Two whitespace indentation, please.

Similarly below.

**Comment by jianqiao** [comment]
LGTM!

**Comment by jianqiao** [comment]
@hbdeshmukh Hi Harshad, can you rebase the branch so I can help merge this PR.

**Comment by hbdeshmukh** [comment]
Hi @jianqiao 

I have rebased and addressed style concerns by @zuyu. Thanks!

**Pull request merged**